app/models/project_topic.rb

```diff
@@ -0,0 +1,95 @@
+class ProjectTopic < ApplicationRecord
+  has_many :signed_up_teams, foreign_key: :sign_up_topic_id, dependent: :destroy
+  has_many :teams, through: :signed_up_teams # list all teams choose this topic, no matter in waitlist or not
+  has_many :assignment_questionnaires, class_name: 'AssignmentQuestionnaire', foreign_key: 'topic_id', dependent: :destroy
+  belongs_to :assignment
+
+  # max_choosers should be a non-negative integer
+  validates :max_choosers, numericality: { only_integer: true, greater_than_or_equal_to: 0 }
+
+  # Checks if there are any slots currently available
+  # Returns true if the number of available slots is greater than 0, otherwise false
+  def slot_available?
+    current_available_slots > 0
+  end
+
+  # Assigns the current topic to a team by updating the provided sign-up record.
+  # Marks the sign-up as not waitlisted and associates it with the current topic.
+  def assign_topic_to_team(new_sign_up)
+    new_sign_up.update(is_waitlisted: false, sign_up_topic_id: self.id)
+  end
+
+  # Adds a new sign-up to the waitlist by updating the sign-up record.
+  # Marks the sign-up as waitlisted and associates it with the current topic.
+  def save_waitlist_entry(new_sign_up)
+    new_sign_up.update(is_waitlisted: true, sign_up_topic_id: self.id)
+  end
+
+  # Signs up a team for the current topic.
+  # Checks if the team is already signed up, and if so, ensures they are not waitlisted.
+  # If a slot is available, assigns the topic to the team; otherwise, adds the team to the waitlist.
+  def sign_up_team(team_id)
+    topic_id = self.id
+
+    # Check if the team has already signed up for this topic
+    existing_sign_up = SignedUpTeam.find_by(sign_up_topic_id: topic_id, team_id: team_id)
+
+    # If the team is already signed up and not waitlisted, return false
+    if !existing_sign_up.nil? && !existing_sign_up.is_waitlisted
+      return false
+    end
+
+    # Create a new sign-up entry for the team
+    new_sign_up = SignedUpTeam.new(sign_up_topic_id: topic_id, team_id: team_id)
+
+     # If there are available slots, assign the topic to the team and remove the team from the waitlist
+    if slot_available?
+      assign_topic_to_team(new_sign_up)
+      result = SignedUpTeam.drop_off_team_waitlists(team_id)
+    else
+      # If no slots are available, add the team to the waitlist
+      result = save_waitlist_entry(new_sign_up)
+    end
+
+    result
+  end
+
+  # Retrieves the team that has been waitlisted the longest for a given topic.
+  # The team is selected based on the earliest created waitlist entry.
```

Comment:
I wouldn't talk about "waitlist entries".  It suggests that waitlists are implemented as actual lists.